### PR TITLE
SONARPY-937 S5655 (ArgumentTypeCheck) should report also on incompati…

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
…ble ambiguous or overloaded functions

### CODE REVIEW

File: python-checks/src/test/resources/checks/argumentType.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def builtin_functions():
   number.__add__(x = unexpected) # Noncompliant {{Change this argument; Function "__add__" expects a different type}}
 #                ^^^^^^^^^^^^^^
   float.fromhex(42) # Noncompliant
-  eval(42)
+  eval(42) # Noncompliant
```
Comment (andrea-guarino-sonarsource) [inline python-checks/src/test/resources/checks/argumentType.py:92]: This is a TP!

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM.
Minor refactoring suggestion & comment on some potential edge case with issue reporting. For that, feel free to see if you prefer to handle it in this PR or just document the behavior.

File: python-checks/src/main/java/org/sonar/python/checks/ArgumentTypeCheck.java [hunk 2]
```diff
@@ -51,46 +59,66 @@ public void initialize(Context context) {
...
+      return false;
+    }
+    boolean isStaticCall = callExpression.callee().is(Tree.Kind.NAME) || Optional.of(callExpression.callee())
+      .filter(c -> c.is(Tree.Kind.QUALIFIED_EXPR))
+      .flatMap(q -> TreeUtils.getSymbolFromTree(((QualifiedExpression) q).qualifier()).filter(s -> s.is(Symbol.Kind.CLASS)))
+      .isPresent();
+    return checkFunctionCall(issue, callExpression, functionSymbol, isStaticCall);
+  }
+
+  private static Set<Symbol> flattenAmbiguousSymbols(Symbol symbol) {
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/ArgumentTypeCheck.java:85]: Good point, thanks! I'll add a helper in `SymbolUtils`

File: python-checks/src/test/resources/checks/argumentType_overloaded_functions.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+overloaded_baz(A())  # OK
+
+
+if predicate():
+  def overloaded_fn(param1: A, param2: A): ...
+else:
+  def overloaded_fn(param1: B, param2: B): ...
+
+overloaded_fn(A(), B())  # Noncompliant
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-checks/src/test/resources/checks/argumentType_overloaded_functions.py]: Very good point, thanks! I finally decided to raise two separate issues (one for each argument), mainly because it was not easy to have a deterministic primary location, as it depends on the order to alternatives in the AmbiguousSymbol.
Let me know if that's fine for you.
Also, actually with `fn` I meant `function` instead of `false negative` 😅  I can change it if it's confusing.

[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply python-checks/src/test/resources/checks/argumentType_overloaded_functions.py]: To be honest I think it's a bit weird to raise multiple issues, as it's really the combination of all argument types that create the issue.
Perhaps we could keep only the first noncompliant argument to appear syntactically for the primary location (by comparing token lines/columns) ? We might also highlight the other arguments for secondary locations.

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM
